### PR TITLE
feat(datasets) Extract Storage classes from Dataset for reuse

### PR DESCRIPTION
**Pull request by fpacifici**
This should be a no brainer. Now that we have a class that represents the Storage in a dataset we can reuse them instead of having weird cross references between datasets.
This moves the Storage objects out of the dataset objects so they can be reused.

Depends on #840

### CODE REVIEW

File: snuba/datasets/storages/events.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+def get_promoted_tags() -> Mapping[str, Sequence[str]]:
+    # The canonical list of foo.bar strings that you can send as a `tags[foo.bar]` query
+    # and they can/will use a promoted column.
+    return {
+        col: [
+            _get_column_tag_map()[col].get(x, x) for x in _get_promoted_columns()[col]
+        ]
+        for col in _get_promoted_columns()
+    }
```
Comment (fpacifici) [inline snuba/datasets/storages/events.py]: These will become irrelevant in https://github.com/getsentry/snuba/pull/774

File: snuba/datasets/storages/events.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    ],
+    order_by="(project_id, toStartOfDay(timestamp), %s)" % sample_expr,
+    partition_by="(toMonday(timestamp), if(equals(retention_days, 30), 30, 90))",
+    version_column="deleted",
+    sample_expr=sample_expr,
+    migration_function=events_migrations,
+)
+
+
+def _get_promoted_columns() -> Mapping[str, FrozenSet[str]]:
```
Comment (fpacifici) [inline snuba/datasets/storages/events.py]: All these methods are moved from the events dataset. 
#774 removes all of these functions.

File: snuba/datasets/storages/events.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+
+def get_promoted_tags() -> Mapping[str, Sequence[str]]:
+    # The canonical list of foo.bar strings that you can send as a `tags[foo.bar]` query
+    # and they can/will use a promoted column.
+    return {
+        col: [get_column_tag_map()[col].get(x, x) for x in get_promoted_columns()[col]]
+        for col in get_promoted_columns()
+    }
```
Comment (fpacifici) [inline snuba/datasets/storages/events.py]: These methods are finally removed in PR #774

**Review by tkaemming — APPROVED** [review]
This makes sense at a high level, mostly relying on tests and type checking for the details due to surface area

**Comment by fpacifici** [comment]
Some context on the testing I did:
- ran snuba automated tests  - ok
- ran sentry automated tests against snuba - ok
- manually ingested events in the following datasets: events, transactions, outcomes, groupedmessages
- manually ran basic queries on each of them + discover and groups (join)

**Pull request merged**